backend/src/main/scala/com/softwaremill/adopttapir/template/BuildView.scala

```diff
@@ -91,18 +93,18 @@ abstract class BuildView:
         List(ScalaTestDependency("com.softwaremill.sttp.client3", "zio-json", TemplateDependencyInfo.sttpVersion))
     }
 
-  private def getHttpDependencies(starterDetails: StarterDetails): List[Dependency] =
+  private def getHttpDependencies(starterDetails: StarterDetails): Either[UnsupportedOperationException, List[Dependency]] =
     starterDetails match {
-      case ServerEffectAndImplementation(FutureEffect, Netty) => HttpDependencies.netty()
-      case ServerEffectAndImplementation(FutureEffect, VertX) => HttpDependencies.vertX()
-      case ServerEffectAndImplementation(IOEffect, Http4s)    => HttpDependencies.http4s()
-      case ServerEffectAndImplementation(IOEffect, Netty)     => HttpDependencies.ioNetty()
-      case ServerEffectAndImplementation(IOEffect, VertX)     => HttpDependencies.ioVerteX()
-      case ServerEffectAndImplementation(ZIOEffect, Http4s)   => HttpDependencies.http4sZIO()
-      case ServerEffectAndImplementation(ZIOEffect, ZIOHttp)  => HttpDependencies.ZIOHttp()
-      case ServerEffectAndImplementation(ZIOEffect, VertX)    => HttpDependencies.ZIOVerteX()
-      case ServerEffectAndImplementation(ZIOEffect, Netty)    => HttpDependencies.ZIONetty()
-      case other: StarterDetails => throw new UnsupportedOperationException(s"Cannot pick dependencies for $other")
+      case ServerEffectAndImplementation(FutureEffect, Netty) => HttpDependencies.netty().asRight
+      case ServerEffectAndImplementation(FutureEffect, VertX) => HttpDependencies.vertX().asRight
+      case ServerEffectAndImplementation(IOEffect, Http4s)    => HttpDependencies.http4s().asRight
+      case ServerEffectAndImplementation(IOEffect, Netty)     => HttpDependencies.ioNetty().asRight
+      case ServerEffectAndImplementation(IOEffect, VertX)     => HttpDependencies.ioVerteX().asRight
+      case ServerEffectAndImplementation(ZIOEffect, Http4s)   => HttpDependencies.http4sZIO().asRight
+      case ServerEffectAndImplementation(ZIOEffect, ZIOHttp)  => HttpDependencies.ZIOHttp().asRight
+      case ServerEffectAndImplementation(ZIOEffect, VertX)    => HttpDependencies.ZIOVerteX().asRight
+      case ServerEffectAndImplementation(ZIOEffect, Netty)    => HttpDependencies.ZIONetty().asRight
+      case other: StarterDetails => UnsupportedOperationException(s"Cannot pick dependencies for $other").asLeft
```

Comment:
maybe we could get rid of the `Either` (which propagates quite deeply), by making sure that we can't create incorrect `StarterDetails` in the first place?

Reply:
Then in the `other` branch we could safely throw an exception as this would be a programming bug, not  a user error

Reply:
Wouldn't it be unsafe to throw an error outside the effect system? This way (with Either converted to IO), it's passed to IO and only kills a fiber. So maybe I could just return `IO` from here.  My idea was to return `Either` to point out this is pure, not side-effecting code, but it can fail. 

So I could move this part of the validation to FormValidator, but then I think it would still be unsafe to throw an exception here (unless I wrap the whole function in `IO.apply` as it was done before).

I was also thinking to change this part to use `PartialFunction`, but I don't think it will be less bloated than Either.

Reply:
How about representing every valid combination of a server and an effect as a value in a sealed ADT, so that illegal combinations are simply not representable? Then the `case other` won't be necessary at all.